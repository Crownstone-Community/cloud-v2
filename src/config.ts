/**
 * This file will house all used environmental variables we use in the application.
 * It does not cover environmental variables used in dependencies.
 */
export const CONFIG = {
  mongoFilesURL: process.env.FILES_DB_URL ?? "mongodb://127.0.0.1:27017/files_v1",
  mongoDataURL:  process.env.DATA_DB_URL  ?? "mongodb://127.0.0.1:27017/data_v1",
  mongoUserURL:  process.env.USER_DB_URL  ?? "mongodb://127.0.0.1:27017/users_v1",

  emailValidationRequired: process.env.EMAIL_VALIDATION_REQUIRED ?? true,
  generateCustomIds: process.env.GENERATE_CUSTOM_IDS ?? false,

  SSEToken: process.env.SSE_TOKEN || "sse_token"
}

export const SecurityTypes = {
  accessToken: 'AccessToken',
}

