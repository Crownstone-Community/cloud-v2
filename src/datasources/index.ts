export * from './data.datasource';
export * from './users.datasource';
export * from './files.datasource';