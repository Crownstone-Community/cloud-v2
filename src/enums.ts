export const keyTypes = {
    ADMIN_KEY:            "ADMIN_KEY",
    MEMBER_KEY:           "MEMBER_KEY",
    BASIC_KEY:            "BASIC_KEY",
    LOCALIZATION_KEY:     "LOCALIZATION_KEY",
    SERVICE_DATA_KEY:     "SERVICE_DATA_KEY",
    MESH_NETWORK_KEY:     "MESH_NETWORK_KEY",
    MESH_APPLICATION_KEY: "MESH_APPLICATION_KEY",
    MESH_DEVICE_KEY:      "MESH_DEVICE_KEY",
    DEVICE_UART_KEY:      "UART_DEVICE_KEY",
};

export const sphereFeatures : Record<string, SphereFeature_t> = {
    ENERGY_COLLECTION_PERMISSION: "ENERGY_COLLECTION_PERMISSION",
}
