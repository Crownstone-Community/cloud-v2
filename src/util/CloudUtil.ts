import * as crypto from "crypto";
const uuid = require('node-uuid');


export const CloudUtil = {
  createKey: function() : string {
    return crypto.randomBytes(16).toString('hex');
  },

  createIBeaconUUID() : string {
    return uuid.v4();
  },

  createToken: function() : string {
    return crypto.randomBytes(32).toString('hex');
  },

  hashPassword(plaintextPassword: string) : string {
    let shasum = crypto.createHash('sha1');
    shasum.update(String(plaintextPassword));
    let hashedPassword = shasum.digest('hex');
    return hashedPassword;
  }
}