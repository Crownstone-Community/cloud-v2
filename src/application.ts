import {BootMixin} from '@loopback/boot';
import {ApplicationConfig} from '@loopback/core';
import {
  RestExplorerBindings,
  RestExplorerComponent,
} from '@loopback/rest-explorer';
import {RepositoryMixin} from '@loopback/repository';
import {RestApplication} from '@loopback/rest';
import {ServiceMixin} from '@loopback/service-proxy';
import path from 'path';
import {CrownstoneSequence} from './sequence';
import {
  AuthenticationComponent,
  registerAuthenticationStrategy
} from '@loopback/authentication';
import {
  AuthorizationComponent,
} from '@loopback/authorization';
import {UserService} from './services';
import {AccessTokenStrategy} from "./security/authentication-strategies/access-token-strategy";
import {SecurityTypes} from "./config";
import {SphereAuthorizationComponent} from "./security/authorization-strategies/authorization";

export interface PackageInfo {
  name: string;
  version: string;
  description: string;
}
const pkg: PackageInfo = require('../package.json');

export class CrownstoneCloud extends BootMixin(ServiceMixin(RepositoryMixin(RestApplication))) {
  constructor(options: ApplicationConfig = {}) {
    let executionPath = __dirname;
    if (options.customPath !== undefined) { executionPath = options.customPath; }
    let customPort = process.env.PORT || 3050;
    if (options.rest && options.rest.port !== undefined) {
      customPort = options.rest.port;
    }

    let customHost = process.env.HOST || '127.0.0.1';
    if (options.rest && options.rest.host !== undefined) {
      customHost = options.rest.host;
    }

    super({...options, rest: { ...options.rest, port: customPort, host: customHost }})

    this.api({
      openapi: '3.0.0',
      info: {title: pkg.name, version: pkg.version},
      paths: {},
      components: {securitySchemes: {
        [SecurityTypes.accessToken]: {
          type: 'apiKey',
          in: 'header',
          name:'Authorization'
        },
      }},
      servers:  [{url: '/api'}],
      security: [{[SecurityTypes.accessToken]: []}],
    });


    this.setUpBindings();
    // Bind authentication component related elements
    this.component(AuthenticationComponent);
    this.component(AuthorizationComponent);
    this.component(SphereAuthorizationComponent);


    // authentication
    registerAuthenticationStrategy(this, AccessTokenStrategy);

    // Set up the custom sequence
    this.sequence(CrownstoneSequence);

    // Set up default home page
    this.static('/', path.join(executionPath, '../public'));

    // Customize @loopback/rest-explorer configuration here
    this.configure(RestExplorerBindings.COMPONENT).to({ path: '/explorer' });
    this.component(RestExplorerComponent);

    this.projectRoot = executionPath;

    // We define both here to allow the testing framework to generate coverage over the ts files.
    this.bootOptions = {
      controllers: {
        dirs: ['controllers'],
        extensions: ['.controller.ts','.controller.js'],
        nested: true,
      },
      repositories: {
        dirs: ['repositories'],
        extensions: ['.repository.ts','.repository.js'],
        nested: true,
      },
      datasources: {
        dirs: ['datasources'],
        extensions: ['.datasource.ts','.datasource.js'],
        nested: true,
      },
      services: {
        dirs: ['services'],
        extensions: ['.service.ts','.service.js'],
        nested: true,
      },
    };
  }

  setUpBindings(): void {
    this.bind("UserService").toClass(UserService);
  }
}
