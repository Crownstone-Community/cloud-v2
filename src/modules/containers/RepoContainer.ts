import {SphereRepository}               from "../../repositories/data/sphere.repository";
import {AppInstallationRepository}      from "../../repositories/data/app-installation.repository";
import {DevicePreferencesRepository}    from "../../repositories/data/device-preferences.repository";
import {DeviceRepository}               from "../../repositories/data/device.repository";
import {FingerprintLinkerRepository}    from "../../repositories/data/fingerprint-linker.repository";
import {FingerprintRepository}          from "../../repositories/data/fingerprint.repository";
import {LocationRepository}             from "../../repositories/data/location.repository";
import {MessageRepository}              from "../../repositories/data/message.repository";
import {MessageStateRepository}         from "../../repositories/data/message-state.repository";
import {MessageUserRepository}          from "../../repositories/data/message-user.repository";
import {SceneRepository}                from "../../repositories/data/scene.repository";
import {SphereAccessRepository}         from "../../repositories/data/sphere-access.repository";
import {SphereFeatureRepository}        from "../../repositories/data/sphere-feature.repository";
import {SphereTrackingNumberRepository} from "../../repositories/data/sphere-tracking-number.repository";
import {SphereKeyRepository}            from "../../repositories/data/sphere-key.repository";
import {StoneRepository}                from "../../repositories/data/stone.repository";
import {StoneAbilityPropertyRepository} from "../../repositories/data/stone-ability-property.repository";
import {StoneAbilityRepository}         from "../../repositories/data/stone-ability.repository";
import {StoneBehaviourRepository}       from "../../repositories/data/stone-behaviour.repository";
import {StoneSwitchStateRepository}     from "../../repositories/data/stone-switch-state.repository";
import {StoneKeyRepository}             from "../../repositories/data/stone-key.repository";
import {ToonRepository}                 from "../../repositories/data/toon.repository";
import {UserRepository}                 from "../../repositories/users/user.repository";
import {HubRepository}                  from "../../repositories/users/hub.repository";
import {CrownstoneTokenRepository}      from "../../repositories/users/crownstone-token.repository";
import {CrownstoneCloud}                from "../../application";
import {FirmwareRepository}             from "../../repositories/data/firmware.repository";
import {BootloaderRepository}           from "../../repositories/data/bootloader.repository";
import {OauthTokenRepository}           from "../../repositories/users/oauth-token.repository";
import {DeviceLocationMapRepository}    from "../../repositories/data/device-location-map.repository";
import {DeviceSphereMapRepository}      from "../../repositories/data/device-sphere-map.repository";
import {FsChunksRepository}             from "../../repositories/files/fs.chunks.repository";
import {FsFilesRepository}              from "../../repositories/files/fs.files.repository";
import { FingerprintV2Repository }      from "../../repositories/data/fingerprint-v2.repository";
import {MessageV2Repository}            from "../../repositories/data/messageV2.repository";
import {MessageRecipientUserRepository} from "../../repositories/data/message-recipient-user.repository";
import {MessageDeletedByUserRepository} from "../../repositories/data/message-deletedBy-user.repository";
import {MessageReadByUserRepository}    from "../../repositories/data/message-readBy-user.repository";
import {AppRepository}                  from "../../repositories/data/app.repository";
import {EnergyDataRepository}           from "../../repositories/data/stone-energy-data.repository";
import {EnergyDataProcessedRepository}  from "../../repositories/data/stone-energy-data-processed.repository";
import {EnergyMetaDataRepository}       from "../../repositories/data/stone-energy-metadata.repository";
import {MetaDataRepository}             from "../../repositories/data/metadata.repository";

export interface RepositoryContainer {
  app:                  AppRepository,
  appInstallation:      AppInstallationRepository,
  bootloader:           BootloaderRepository,
  oauthToken:           OauthTokenRepository,
  crownstoneToken:      CrownstoneTokenRepository,
  devicePreferences:    DevicePreferencesRepository,
  deviceLocationMap:    DeviceLocationMapRepository,
  deviceSphereMap:      DeviceSphereMapRepository,
  device:               DeviceRepository,
  fingerprintLinker:    FingerprintLinkerRepository,
  fingerprint:          FingerprintRepository,
  fingerprintV2:        FingerprintV2Repository,
  firmware:             FirmwareRepository,
  fsChunks:             FsChunksRepository,
  fsFiles:              FsFilesRepository,
  hub:                  HubRepository,
  location:             LocationRepository,
  messageV2:            MessageV2Repository,
  messageRecipientUser: MessageRecipientUserRepository,
  messageReadByUser:    MessageReadByUserRepository,
  messageDeletedByUser: MessageDeletedByUserRepository,
  message:              MessageRepository,
  messageState:         MessageStateRepository,
  messageUser:          MessageUserRepository,
  metaData:             MetaDataRepository,
  scene:                SceneRepository,
  sphereAccess:         SphereAccessRepository,
  sphereFeature:        SphereFeatureRepository,
  sphereTrackingNumber: SphereTrackingNumberRepository,
  sphereKeys:           SphereKeyRepository,
  sphere:               SphereRepository,
  stone:                StoneRepository,
  stoneAbilityProperty: StoneAbilityPropertyRepository,
  stoneAbility:         StoneAbilityRepository,
  stoneBehaviour:       StoneBehaviourRepository,
  stoneSwitchState:     StoneSwitchStateRepository,
  stoneKeys:            StoneKeyRepository,
  stoneEnergy:          EnergyDataRepository;
  stoneEnergyProcessed: EnergyDataProcessedRepository;
  stoneEnergyMetaData:  EnergyMetaDataRepository;
  toon:                 ToonRepository,
  user:                 UserRepository,
}


export let Dbs : RepositoryContainer = {
  app:                  null,
  appInstallation:      null,
  bootloader:           null,
  oauthToken:           null,
  crownstoneToken:      null,
  deviceLocationMap:    null,
  deviceSphereMap:      null,
  devicePreferences:    null,
  device:               null,
  fsChunks:             null,
  fsFiles:              null,
  fingerprintLinker:    null,
  fingerprint:          null,
  fingerprintV2:        null,
  firmware:             null,
  hub:                  null,
  location:             null,
  messageV2:            null,
  messageRecipientUser: null,
  messageReadByUser:    null,
  messageDeletedByUser: null,
  message:              null,
  messageState:         null,
  messageUser:          null,
  metaData:             null,
  scene:                null,
  sphereAccess:         null,
  sphereFeature:        null,
  sphereTrackingNumber: null,
  sphereKeys:           null,
  sphere:               null,
  stone:                null,
  stoneAbilityProperty: null,
  stoneAbility:         null,
  stoneBehaviour:       null,
  stoneSwitchState:     null,
  stoneKeys:            null,
  stoneEnergy:          null,
  stoneEnergyProcessed: null,
  stoneEnergyMetaData:  null,
  toon:                 null,
  user:                 null,
};

export async function PopulateRepositoryContainer(app: CrownstoneCloud) {
  Dbs.app                  = await app.getRepository( AppRepository );
  Dbs.appInstallation      = await app.getRepository( AppInstallationRepository );
  Dbs.bootloader           = await app.getRepository( BootloaderRepository );
  Dbs.oauthToken           = await app.getRepository( OauthTokenRepository );
  Dbs.crownstoneToken      = await app.getRepository( CrownstoneTokenRepository );
  Dbs.deviceLocationMap    = await app.getRepository( DeviceLocationMapRepository );
  Dbs.deviceSphereMap      = await app.getRepository( DeviceSphereMapRepository );
  Dbs.devicePreferences    = await app.getRepository( DevicePreferencesRepository );
  Dbs.device               = await app.getRepository( DeviceRepository );
  Dbs.fingerprintLinker    = await app.getRepository( FingerprintLinkerRepository );
  Dbs.fingerprint          = await app.getRepository( FingerprintRepository );
  Dbs.fingerprintV2        = await app.getRepository( FingerprintV2Repository );
  Dbs.firmware             = await app.getRepository( FirmwareRepository );
  Dbs.fsChunks             = await app.getRepository( FsChunksRepository );
  Dbs.fsFiles              = await app.getRepository( FsFilesRepository );
  Dbs.hub                  = await app.getRepository( HubRepository );
  Dbs.location             = await app.getRepository( LocationRepository );
  Dbs.messageV2            = await app.getRepository( MessageV2Repository );
  Dbs.messageRecipientUser = await app.getRepository( MessageRecipientUserRepository );
  Dbs.messageReadByUser    = await app.getRepository( MessageReadByUserRepository );
  Dbs.messageDeletedByUser = await app.getRepository( MessageDeletedByUserRepository );
  Dbs.message              = await app.getRepository( MessageRepository );
  Dbs.messageState         = await app.getRepository( MessageStateRepository );
  Dbs.messageUser          = await app.getRepository( MessageUserRepository );
  Dbs.metaData             = await app.getRepository( MetaDataRepository );
  Dbs.scene                = await app.getRepository( SceneRepository );
  Dbs.sphereAccess         = await app.getRepository( SphereAccessRepository );
  Dbs.sphereFeature        = await app.getRepository( SphereFeatureRepository );
  Dbs.sphereTrackingNumber = await app.getRepository( SphereTrackingNumberRepository );
  Dbs.sphereKeys           = await app.getRepository( SphereKeyRepository );
  Dbs.sphere               = await app.getRepository( SphereRepository );
  Dbs.stone                = await app.getRepository( StoneRepository );
  Dbs.stoneAbilityProperty = await app.getRepository( StoneAbilityPropertyRepository );
  Dbs.stoneAbility         = await app.getRepository( StoneAbilityRepository );
  Dbs.stoneBehaviour       = await app.getRepository( StoneBehaviourRepository );
  Dbs.stoneSwitchState     = await app.getRepository( StoneSwitchStateRepository );
  Dbs.stoneKeys            = await app.getRepository( StoneKeyRepository );
  Dbs.stoneEnergy          = await app.getRepository( EnergyDataRepository );
  Dbs.stoneEnergyMetaData  = await app.getRepository( EnergyMetaDataRepository );
  Dbs.stoneEnergyProcessed = await app.getRepository( EnergyDataProcessedRepository );
  Dbs.toon                 = await app.getRepository( ToonRepository );
  Dbs.user                 = await app.getRepository( UserRepository );
}
