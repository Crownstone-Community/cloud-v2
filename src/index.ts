import {ApplicationConfig, ExpressServer} from './server';
import {PopulateRepositoryContainer} from "./modules/containers/RepoContainer";

export * from './server';

const config = {
  rest: {
    // Use the LB4 application as a route. It should not be listening.
    listenOnStart: false,
  },
};

export async function main(options: ApplicationConfig = {}) {

  const server = new ExpressServer(config);
  await server.boot();
  await server.start();

  await PopulateRepositoryContainer(server.lbApp);

  const port = server.lbApp.restServer.config.port ?? 3000;
  const host = server.lbApp.restServer.config.host ?? 'NO-HOST';

  console.log(`Server is running at ${host}:${port}`);

  return server.lbApp;
}

if (require.main === module) {
  // Run the application
  main({}).catch(err => {
    console.error('Cannot start the application.', err);
    process.exit(1);
  });
}
