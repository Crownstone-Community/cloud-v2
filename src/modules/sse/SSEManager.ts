import { Server, Socket } from "socket.io"
import * as http from "http";
import Timeout = NodeJS.Timeout;
import * as crypto from "crypto";
import {Dbs} from "../containers/RepoContainer";
let config  = require('../config.' + (process.env.NODE_ENV || 'local'));

const S4 = function () {
  return Math.floor(Math.random() * 0x10000 /* 65536 */).toString(16);
};

const getShortUUID = function() {
  return (
    S4() + S4() + '-' +
    S4()
  );
}

const EVENT_ROOM_NAME = "/SSE_EVENTS";

const protocolTopics = {
  requestForOauthTokenCheck:  "requestForOauthTokenCheck",
  requestForAccessTokenCheck: "requestForAccessTokenCheck",
  authenticationRequest:      "authenticationRequest",
  event:                      "event",
}



class SSEManagerClass {
  io : Server;
  connections : {[id: string]: SSEConnection} = {};

  constructor() {
    this.io = null;
    this.connections = {};
  }

  init(server: http.Server) {
    this.io = new Server(server, { pingInterval: 4000, pingTimeout: 2000, transports:["websocket"], cookie:false })
    this.io.on('connect', (socket: Socket) => {
      let uid = getShortUUID()
      this.connections[uid] = new SSEConnection(socket, () => { delete this.connections[uid]; });
    });
  }

  emit(data: SseEvent) {
    this.io.sockets.in(EVENT_ROOM_NAME).emit(protocolTopics.event, data);
  }
}

export const SSEManager = new SSEManagerClass();


class SSEConnection {

  socket : Socket
  cleanup : () => void;
  authenticationTimeout : Timeout;
  handshake : string;

  constructor(socket: Socket, cleanup: () => void) {
    this.socket = null;
    this.cleanup = null;
    this.authenticationTimeout = null;
    this.handshake = null;

    this.handshake = crypto.randomBytes(16).toString('base64')

    this.socket = socket;
    this.cleanup = cleanup;

    // client gets 500 ms to authenticate
    this.authenticationTimeout = setTimeout(() => { this.destroy(); }, 500);

    // here we ensure only our SSE servers will connect to this socket before sending data.
    this.socket.emit(protocolTopics.authenticationRequest, this.handshake, (reply: string) => {
      if (this.authenticate(reply) === false) { this.destroy(); return; }
      clearTimeout(this.authenticationTimeout);

      // add to the list of sockets that can process SSE's
      this.socket.join(EVENT_ROOM_NAME);

      // check if an accesstoken is valid.
      this.socket.on(protocolTopics.requestForAccessTokenCheck, (token, callback) => {
        this.handleAccessTokenRequest(token, callback);
      })
      // check if an accesstoken is valid.
      this.socket.on(protocolTopics.requestForOauthTokenCheck, (token, callback) => {
        this.handleOauthTokenRequest(token, callback);
      })
    });

    this.socket.on("disconnect", () => { this.destroy(); });
  }


  destroy() {
    this.socket.disconnect(true);
    clearTimeout(this.authenticationTimeout);
    this.cleanup();
  }


  authenticate(reply: string) {
    let hasher = crypto.createHash('sha256');
    let output = hasher.update(this.handshake + config.SSEToken).digest('hex')
    return reply === output;
  }


  async _isValidAccessToken(token: string) : Promise<AccessModel> {
    let resultTokenData: AccessModel = {
      accessToken: token,
      ttl: 0,
      createdAt: new Date(),
      userId: null,
      spheres: {},
      scopes: ["all"],
    };

    // get the token from the db, if successful,
    let tokenData = await Dbs.crownstoneToken.findById(token);
    if (!tokenData) { throw "INVALID_TOKEN" }
    if (new Date().valueOf() < new Date(tokenData.created).valueOf() + tokenData.ttl*1000) {
      throw "EXPIRED_TOKEN"
    }
    resultTokenData.ttl = tokenData.ttl;
    resultTokenData.createdAt = tokenData.created;

    resultTokenData.userId = tokenData.userId;

    let access = await Dbs.sphereAccess.find({where: {and: [{userId: tokenData.userId}, {invitePending: {neq: true}}]}, fields: {sphereId: true}})

    for (let i = 0; i < access.length; i++) {
      resultTokenData.spheres[access[i].sphereId] = true;
    }
    return resultTokenData;
  }

  async _isValidOauthToken(token: string) : Promise<AccessModel> {
    let resultTokenData: AccessModel = {
      accessToken: token,
      ttl: 0,
      createdAt: new Date(),
      userId: null,
      spheres: {},
      scopes: [],
    };

    // get the token from the db, if successful,
    let tokenData = await Dbs.oauthToken.findById(token);
    if (!tokenData) { throw "INVALID_TOKEN" }
    if (new Date().valueOf() < new Date(tokenData.expiredAt).valueOf()) {
      throw "EXPIRED_TOKEN"
    }
    resultTokenData.ttl       = tokenData.expiresIn;
    resultTokenData.createdAt = tokenData.issuedAt;
    resultTokenData.userId    = tokenData.userId;
    resultTokenData.scopes    = tokenData.scopes as oauthScope[];

    let access = await Dbs.sphereAccess.find({where: {and: [{userId: tokenData.userId}, {invitePending: {neq: true}}]}, fields: {sphereId: true}})

    for (let i = 0; i < access.length; i++) {
      resultTokenData.spheres[access[i].sphereId] = true;
    }
    return resultTokenData;
  }

  handleAccessTokenRequest(token: string, callback: (result: any) => void) {
    this._isValidAccessToken(token)
      .then((result) => {
        if (!result) { throw "Invalid Token" }

        callback({code: 200, data: result});
      })
      .catch((err) => {
        return callback({code: 401, err: err});
      })
  }

  handleOauthTokenRequest(token: string, callback: (result: any) => void) {
    this._isValidOauthToken(token)
      .then((result) => {
        if (!result) { throw "Invalid Token" }

        callback({code: 200, data: result});
      })
      .catch((err) => {
        return callback({code: 401, err: err});
      })
  }
}
