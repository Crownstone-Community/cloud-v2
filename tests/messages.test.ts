import { CONFIG } from "../src/config";
CONFIG.emailValidationRequired = false;
CONFIG.generateCustomIds = true;

import { mockSSEmanager } from "./mocks/SSE";
mockSSEmanager();
import { SSEManager } from "../src/modules/sse/SSEManager";

import {
  makeUtilDeterministic,
  resetMockDatabaseIds,
  resetMockRandom,
  setDate,
} from "./mocks/CloudUtil.mock";
makeUtilDeterministic();

import { CrownstoneCloud } from "../src/application";
import { Client, createRestAppClient } from "@loopback/testlab";
import { clearTestDatabase, createApp, getRepositories } from "./helpers";
import {
  createHub,
  createLocation,
  createSphere,
  createStone,
  createUser,
  resetUsers,
} from "./builders/createUserData";
import { auth, getToken } from "./rest-helpers/rest.helpers";
import { CloudUtil } from "../src/util/CloudUtil";
import { mocked } from "ts-jest/utils";
import { Dbs } from "../src/modules/containers/RepoContainer";

let app: CrownstoneCloud;
let client: Client;

let dbs;
let admin;
let member;
let guest;
let sphere;
let hub;
let stone, behaviour, ability, abilityProperty;
let stone2, behaviour2, ability2, abilityProperty2;
let stone3, behaviour3, ability3, abilityProperty3;
let location;
let token;

async function populate() {
  // fill with a bit of data for sync
  dbs = getRepositories();
  admin = await createUser("test@test.com", "test", 0);
  member = await createUser("member@test.com", "test", 0);
  guest = await createUser("guest@test.com", "test", 0);
  sphere = await createSphere(admin.id, "mySphere", 0);
  await dbs.sphereAccess.create({
    sphereId: sphere.id,
    userId: member.id,
    role: "member",
    sphereAuthorizationToken: CloudUtil.createToken(),
    invitePending: false,
  });
  await dbs.sphereAccess.create({
    sphereId: sphere.id,
    userId: guest.id,
    role: "guest",
    sphereAuthorizationToken: CloudUtil.createToken(),
  });
  hub = await createHub(sphere.id, "myHub", 0);
  ({ stone, behaviour, ability, abilityProperty } = await createStone(
    sphere.id,
    "stone1",
    0
  ));
  ({
    stone: stone2,
    behaviour: behaviour2,
    ability: ability2,
    abilityProperty: abilityProperty2,
  } = await createStone(sphere.id, "stone2", 0));
  ({
    stone: stone3,
    behaviour: behaviour3,
    ability: ability3,
    abilityProperty: abilityProperty3,
  } = await createStone(sphere.id, "stone3", 0));
  location = await createLocation(sphere.id, "location", 0);

  stone.locationId = location.id;
  await dbs.stone.update(stone);

  // sets the token for the admin user for the auth function.
  token = await getToken(client, admin);
}

beforeEach(async () => {
  mocked(SSEManager.emit).mockReset();
  await clearTestDatabase();
  resetUsers();
  resetMockRandom();
  resetMockDatabaseIds();
});

beforeAll(async () => {
  app = await createApp();
  client = createRestAppClient(app);
});
afterAll(async () => {
  await app.stop();
});

test("Create and get messages", async () => {
  await populate();
  await client.post(auth(`/spheres/${sphere.id}/message`)).send({
    message: {
      content: "member&guest:Message", everyoneInSphere: false, everyoneInSphereIncludingOwner: false},
      recipients: [member.id, guest.id],
  });
  await client.post(auth(`/spheres/${sphere.id}/message`)).send({
    message: {
      content: "guest:Message", everyoneInSphere: false, everyoneInSphereIncludingOwner: false},
      recipients: [guest.id],
  });

  await client.post(auth(`/spheres/${sphere.id}/message`)).send({
    message: {
      content: "everyoneMessageIncAdmin", everyoneInSphere: true, everyoneInSphereIncludingOwner: true},
  });

  await client.post(auth(`/spheres/${sphere.id}/message`)).send({
    message: {
      content: "everyoneMessageExOwner", everyoneInSphere: true, everyoneInSphereIncludingOwner: false},
  });

  await client
    .get(auth(`/spheres/${sphere.id}/messages`))
    .expect(({ body }) => {
      expect(body).toMatchSnapshot('Admins messages');
    });

  let memberToken = await getToken(client, member);

  await client
    .get(auth(`/spheres/${sphere.id}/messages`, memberToken))
    .expect(({ body }) => {
      expect(body).toMatchSnapshot('messages for member');
    });

  let guestToken = await getToken(client, guest);

  await client
    .get(auth(`/spheres/${sphere.id}/messages`, guestToken))
    .expect(({ body }) => {
      expect(body).toMatchSnapshot('messages for guest');
    });
});

test("Create and get messages sent to self without duplicates", async () => {
  await populate();
  await client.post(auth(`/spheres/${sphere.id}/message`)).send({
    message: {
      content: "adminDirect:Message", everyoneInSphere: false, everyoneInSphereIncludingOwner: false},
      recipients: [admin.id],
  });

  await client
    .get(auth(`/spheres/${sphere.id}/messages`))
    .expect(({ body }) => {
      expect(body).toHaveLength(1)
      expect(body).toMatchSnapshot('Admins direct messages');
    });
});


test("Create and get messages by the sender", async () => {
  await populate();
  await client.post(auth(`/spheres/${sphere.id}/message`)).send({
    message: {content: "adminMessage", everyoneInSphere: true, everyoneInSphereIncludingOwner: true},
    recipients: [],
  });

  await client
    .get(auth(`/spheres/${sphere.id}/messages`))
    .expect(({ body }) => {
      expect(body.length).toBe(1);  // only the admin message is returned
    });
});


test("Check reading and app-deleting the message", async () => {
  await populate();
  let message;
  await client.post(auth(`/spheres/${sphere.id}/message`)).send({
    message: {
      content: "memberMessage", everyoneInSphere: false, everyoneInSphereIncludingOwner: false},
    recipients: [member.id],
  })
    .expect(({body}) => { message = body; });


  // member marks message gets it

  token = await getToken(client, member);
  await client.post(auth(`/messages/${message.id}/markAsRead`));

  expect(await Dbs.messageReadByUser.find()).toHaveLength(1);
  expect(await Dbs.messageDeletedByUser.find()).toHaveLength(0);

  await client.post(auth(`/messages/${message.id}/markAsDeleted`));

  expect(await Dbs.messageDeletedByUser.find()).toHaveLength(1);

  await client
    .get(auth(`/spheres/${sphere.id}/messages`))
    .expect(({ body }) => {
      expect(body).toMatchSnapshot();
    });

});
