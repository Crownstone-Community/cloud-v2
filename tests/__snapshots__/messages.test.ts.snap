// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check reading and app-deleting the message 1`] = `
Array [
  Object {
    "content": "memberMessage",
    "createdAt": "1970-01-01T00:00:00.000Z",
    "deletedBy": Array [
      Object {
        "id": "dbId:UserRepository:2",
      },
    ],
    "everyoneInSphere": false,
    "everyoneInSphereIncludingOwner": false,
    "id": "dbId:MessageV2Repository:1",
    "ownerId": "dbId:UserRepository:1",
    "readBy": Array [
      Object {
        "id": "dbId:UserRepository:2",
      },
    ],
    "recipients": Array [
      Object {
        "id": "dbId:UserRepository:2",
      },
    ],
    "sphereId": "dbId:SphereRepository:1",
    "triggerEvent": "enter",
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
]
`;

exports[`Create and get messages 1`] = `
Array [
  Object {
    "content": "everyoneMessageIncAdmin",
    "createdAt": "1970-01-01T00:00:00.000Z",
    "everyoneInSphere": true,
    "everyoneInSphereIncludingOwner": true,
    "id": "dbId:MessageV2Repository:3",
    "ownerId": "dbId:UserRepository:1",
    "sphereId": "dbId:SphereRepository:1",
    "triggerEvent": "enter",
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
]
`;

exports[`Create and get messages 2`] = `
Array [
  Object {
    "content": "memberMessage",
    "createdAt": "1970-01-01T00:00:00.000Z",
    "everyoneInSphere": false,
    "everyoneInSphereIncludingOwner": false,
    "id": "dbId:MessageV2Repository:1",
    "ownerId": "dbId:UserRepository:1",
    "recipients": Array [
      Object {
        "id": "dbId:UserRepository:2",
      },
      Object {
        "id": "dbId:UserRepository:3",
      },
    ],
    "sphereId": "dbId:SphereRepository:1",
    "triggerEvent": "enter",
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
  Object {
    "content": "everyoneMessageIncAdmin",
    "createdAt": "1970-01-01T00:00:00.000Z",
    "everyoneInSphere": true,
    "everyoneInSphereIncludingOwner": true,
    "id": "dbId:MessageV2Repository:3",
    "ownerId": "dbId:UserRepository:1",
    "sphereId": "dbId:SphereRepository:1",
    "triggerEvent": "enter",
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
  Object {
    "content": "everyoneMessageExAdmin",
    "createdAt": "1970-01-01T00:00:00.000Z",
    "everyoneInSphere": true,
    "everyoneInSphereIncludingOwner": false,
    "id": "dbId:MessageV2Repository:4",
    "ownerId": "dbId:UserRepository:1",
    "sphereId": "dbId:SphereRepository:1",
    "triggerEvent": "enter",
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
]
`;
